src/chainstate/stacks/index/trie.rs

```diff
@@ -0,0 +1,2053 @@
+            Ok(children_root_hash.clone())
+        }
+    }
+
+    /// Unwind a TrieCursor to update the Merkle root of the trie.
+    /// The root hashes of each trie form a Merkle skip-list -- the hash of Trie i is calculated
+    /// from the hash of its children, plus the hash Tries i-1, i-2, i-4, i-8, ..., i-2**j, ...
+    /// This is required for Merkle proofs to work (specifically, the shunt proofs).
+    fn recalculate_root_hash(storage: &mut TrieFileStorage, cursor: &TrieCursor, update_skiplist: bool) -> Result<(), Error> {
+        assert!(cursor.node_ptrs.len() > 0);
+
+        let mut ptrs = cursor.node_ptrs.clone();
+        trace!("update_root_hash: ptrs = {:?}", &ptrs);
+        let mut child_ptr = ptrs.pop().unwrap();
+
+        if ptrs.len() == 0 {
+            // root node was already updated by trie operations, but it will have the wrong hash.
+            // we need to "fix" the root node so it mixes in its ancestor hashes.
+            trace!("Fix up root node so it mixes in its ancestor hashes");
+            let (node, cur_hash) = storage.read_nodetype(&child_ptr)?;
+            if node.is_node256() {
+                let root_disk_ptr = storage.root_ptr();
+                let root_ptr = TriePtr::new(TrieNodeID::Node256, 0, root_disk_ptr as u32);
+                if child_ptr != root_ptr {
+                    return Err(Error::CorruptionError("Only ptr is not the root".to_string()));
+                }
+                
+                let mut hash_buf = Vec::with_capacity(TRIEHASH_ENCODED_SIZE * 256);
+                Trie::get_children_hashes_bytes(storage, &node, &mut hash_buf)?;
+
+                let h = 
+                    if update_skiplist {
+                        trace!("Update root skiplist");
+                        Trie::get_trie_root_hash(storage, &get_nodetype_hash_bytes(&node, &hash_buf))?
+                    }
+                    else {
+                        trace!("Not updating root skiplist");
+                        get_nodetype_hash_bytes(&node, &hash_buf)
+                    };
+
+                // for debug purposes
+                if is_trace() {
+                    let node_hash = get_nodetype_hash_bytes(&node, &hash_buf);
+                    let hs = Trie::get_trie_root_ancestor_hashes(storage, &node_hash)?;
+                    trace!("update_root_hash: Updated {:?} with {:?} from {:?} to {:?} + {:?} = {:?} (fixed root)", &node, &child_ptr, &cur_hash, &node_hash, &hs[1..].to_vec(), &h);
+                }
+
+                storage.write_nodetype(child_ptr.ptr(), &node, h)?;
+            }
+            else {
+                return Err(Error::CorruptionError("Only ptr was not a node256".to_string()));
+            }
+        }
+        else {
+            while ptrs.len() > 0 {
+                let ptr = match ptrs.pop() {
+                    Some(p) => p,
+                    None => panic!("Out of ptrs")
+                };
+                if is_backptr(ptr.id()) {
```

Comment:
Is this possible? If the hash needs to be updated from a cursor, then any back pointers should have been copied, and then the node_ptrs will contain the new (non-back-pointing) node pointers. If so, this should panic because otherwise, this path means something has gone wrong.

Reply:
Yeah, I think this is an artifact from an earlier implementation of the TrieCursor where backptrs could be stored in the list of ptrs.  Will change this to an `assert!`